src/test/unit/io/deserializer_test.cpp

```diff
@@ -0,0 +1,899 @@
+  stan::io::deserializer<double> deserializer(theta, theta_i);
+  for (size_t i = 0; i < 7U; ++i) {
+    double x = deserializer.read<double>();
+    EXPECT_FLOAT_EQ(static_cast<double>(i), x);
+  }
+  using complex_array = std::vector<std::complex<double>>;
+  complex_array y = deserializer.read<complex_array>(4);
+  EXPECT_EQ(4, y.size());
+  double sentinal = 7;
+  for (int i = 0; i < y.size(); ++i) {
+    EXPECT_FLOAT_EQ(sentinal, y[i].real());
+    ++sentinal;
+    EXPECT_FLOAT_EQ(sentinal, y[i].imag());
+    ++sentinal;
+  }
+
+  double z = deserializer.read<double>();
+  EXPECT_FLOAT_EQ(15.0, z);
+}
+
+TEST(deserializer_array, read_vector) {
+  std::vector<int> theta_i;
+  std::vector<double> theta;
+  for (size_t i = 0; i < 100U; ++i)
+    theta.push_back(static_cast<double>(i));
+  stan::io::deserializer<double> deserializer(theta, theta_i);
+  std::vector<Eigen::VectorXd> y
+      = deserializer.read<std::vector<Eigen::VectorXd>>(4, 2);
+  EXPECT_EQ(4, y.size());
+  EXPECT_FLOAT_EQ(0.0, y[0](0));
+  EXPECT_FLOAT_EQ(1.0, y[0](1));
+  EXPECT_FLOAT_EQ(2.0, y[1](0));
+  EXPECT_FLOAT_EQ(3.0, y[1](1));
+  EXPECT_FLOAT_EQ(4.0, y[2](0));
+  EXPECT_FLOAT_EQ(5.0, y[2](1));
+  EXPECT_FLOAT_EQ(6.0, y[3](0));
+  EXPECT_FLOAT_EQ(7.0, y[3](1));
+
+  double z = deserializer.read<double>();
+  EXPECT_FLOAT_EQ(8.0, z);
+}
+
+// size zero
+
+TEST(deserializer, zeroSizeVecs) {
+  std::vector<int> theta_i;
+  std::vector<double> theta;
+  theta.push_back(1.0);
+  stan::io::deserializer<double> deserializer(theta, theta_i);
+
+  EXPECT_FLOAT_EQ(1.0, deserializer.read<double>());  // finish available
+
+  EXPECT_EQ(0, deserializer.read<std::vector<double>>(0).size());
+  EXPECT_EQ(0, deserializer.read<Eigen::VectorXd>(0).size());
+  EXPECT_EQ(0, deserializer.read<Eigen::RowVectorXd>(0).size());
+  EXPECT_EQ(0, deserializer.read<Eigen::MatrixXd>(0, 3).size());
+  EXPECT_EQ(0, deserializer.read<Eigen::MatrixXd>(3, 0).size());
+  EXPECT_EQ(0, deserializer
+                   .read<std::vector<std::vector<Eigen::MatrixXd>>>(0, 0, 0, 0)
+                   .size());
+}
+
+// out of memory
+
+TEST(deserializer, eos_exception) {
+  std::vector<double> theta;
+  theta.push_back(1.0);
+  theta.push_back(2.0);
+  std::vector<int> theta_i;
+  theta_i.push_back(1);
+  stan::io::deserializer<double> deserializer(theta, theta_i);
+
+  EXPECT_EQ(2U, deserializer.available());
+  EXPECT_EQ(1U, deserializer.available_i());
+
+  EXPECT_NO_THROW(deserializer.read<double>());
+  EXPECT_NO_THROW(deserializer.read<double>());
+  EXPECT_THROW(deserializer.read<double>(), std::runtime_error);
+
+  // should go back to working
+  EXPECT_NO_THROW(deserializer.read<int>());
+  EXPECT_THROW(deserializer.read<int>(), std::runtime_error);
+
+  // should keep throwing
+  EXPECT_THROW(deserializer.read<double>(), std::runtime_error);
+  EXPECT_THROW(deserializer.read<size_t>(), std::runtime_error);
+
+  EXPECT_THROW(deserializer.read<std::complex<double>>(), std::runtime_error);
```

Comment:
@SteveBronder these were the only problems I found working through this. We throw an error if we read a real and don't have memory for it, so I think we should do similarly for the vectors and stuff. Can you add these in?

Reply:
I added checks for this using `check_r_capacity`.